### PR TITLE
CDI-699 AnnotationLiteral should use privileged actions for reflective operations

### PR DESCRIPTION
**Pull request by antoinesd**
Addition of privileged bloc in AnnotationLiteral

### CODE REVIEW

File: api/src/test/java/org/jboss/cdi/api/test/privileged/annotation/AnnotationLiteralTest.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+package org.jboss.cdi.api.test.privileged.annotation;
+
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+/**
+ *
+ * This class test usage of privileged bloc in {@link javax.enterprise.util.AnnotationLiteral}
+ * They launched with Java security Manager.
```
Comment (mkouba) [inline api/src/test/java/org/jboss/cdi/api/test/privileged/annotation/AnnotationLiteralTest.java]: Where do we run this test with a `SecurityManager` used?

Comment (mkouba) [reply api/src/test/java/org/jboss/cdi/api/test/privileged/annotation/AnnotationLiteralTest.java]: Ah, you pushed one commit before this PR... https://github.com/cdi-spec/cdi/commit/c9aedeaadc9828cf2b5b8c3efc37f733a8f7569a

**Comment by mkouba** [comment]
I think you should also remove https://github.com/cdi-spec/cdi/blob/master/api/pom.xml#L301.

**Comment by mkouba** [comment]
Hm, it seems that `AnnotationLiteralTest` cannot reveal the problem with `java.lang.Class.getDeclaredMethods()` because (javadoc):
````
Throws:
SecurityException - If a security manager, s, is present and any of the following conditions is met:
* the caller's class loader is not the same as the class loader of this class and invocation of s.checkPermission method with RuntimePermission("accessDeclaredMembers") denies access to the declared methods within this class
* the caller's class loader is not the same as or an ancestor of the class loader for the current class and invocation of s.checkPackageAccess() denies access to the package of this class
```

**Comment by manovotn** [comment]
The test configuration is really puzzling, I don't see why you firstly skip test execution and then re-enable it in both phases (e.g. [L301](https://github.com/cdi-spec/cdi/blob/master/api/pom.xml#L301), [L314](https://github.com/cdi-spec/cdi/blob/master/api/pom.xml#L314), [L327](https://github.com/cdi-spec/cdi/blob/master/api/pom.xml#L327)). If you remove all three, I think it will still execute - or is there any special reason for it?

File: api/src/test/java/org/jboss/cdi/api/test/privileged/annotation/AnnotationLiteralTest.java (hunk 1 above)
Comment (antoinesd) [reply api/src/test/java/org/jboss/cdi/api/test/privileged/annotation/AnnotationLiteralTest.java]: Yes I hesitated but I have several PR to send with SM activated so as this addition doesn't change anything to release and test I found more convenient to have this on the common branch.

**Comment by antoinesd** [comment]
@manovotn I need 2 test suites because the standard test are doing a lot of reflection and I couldn't activate SM for them. Adding a second test suite was necessary to activate SM for it. I have problem while overriding the standard test so I have to cancel it (`<skip>true</skip>`). But perhaps I can give another try...

**Comment by antoinesd** [comment]
> The test configuration is really puzzling

I achieved to simplify surefire config and make tests launch clearer by removing `<skip>true</skip>` and override `default-test` suite instead of disabling it and create a new one.

**Comment by struberg** [comment]
Isn't this an impl detail? Ofc it's the official spec api, but this doesn't change the contract. So nothing needs to be released spec wise, right?

**Comment by manovotn** [comment]
> Isn't this an impl detail?

@struberg no it is not. User can have his own implementation of `AnnotationLiteral` which naturally extends this class. And on instantiation of that class the constructor on `AnnotationLiteral` is invoked. That in turn invokes `getMembers()` method which then invokes `getDeclaredMethods()` and that can cause issues with security manager.

**Comment by struberg** [comment]
@manovotn yes, the api.jar should get fixed. But that can be done as a simple maintenance release. It doesn't require a new spec revision.

**Comment by mkouba** [comment]
@struberg Yes, we target a "simple API maintenance release" named _CDI 2.0.SP1_; except that it's probably not so simple because of JCP rules etc. @antoinesd knows more.

**Comment by antoinesd** [comment]
@struberg. It's a maintenance release. Doing it in the context of the JCP would have required to launch a MR (more than 60 days).

**Comment by struberg** [comment]
oki @antoinesd and @mkouba I think we are on the same page. Wenn talking about a cdi-api 'Maintenance Release' I also didn't mean a CDI Spec MR.
In my eyes it's really just a bugfix release of the cdi-api.jar. Kind of 2.0.1. 

In any case nothing which requires to involve the EE EC for getting it to the people.

**Pull request merged**